### PR TITLE
visualize dynamic event sources on topology view

### PR DESCRIPTION
**Pull request by sahil143**
**Fixes**: https://issues.redhat.com/browse/ODC-3404

This PR is dependent on the utility `event-source-dynamic-models.ts`  from this PR https://github.com/openshift/console/pull/4991/

**Analysis / Root cause**: 

There was no way to show dynamic event sources on the topology view

**Solution Description**: 

once a dynamic event source is created it appears on topology with correct sidebar and context menu

**Screen shots / Gifs for design review**: 

Screenshot from 2020-04-13 20-58-31
Screenshot from 2020-04-13 20-58-40
Screenshot from 2020-04-13 20-59-03

**Unit test coverage report**: 

Screenshot from 2020-04-14 18-49-20

**Test setup:**

**Browser conformance**: 

- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/knative-plugin/src/utils/event-source-dynamic-models.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+      } = crd;
+      return {
+        apiGroup: group,
+        apiVersion: version,
+        kind: names?.kind,
+        plural: names?.plural,
+        id: names?.singular,
+        label: names?.singular,
+        labelPlural: names?.plural,
+        abbr: names?.kind?.replace(/[a-z]/g, ''),
```
[the remark this answers is not in the document]
Comment (sahil143) [reply frontend/packages/knative-plugin/src/utils/event-source-dynamic-models.ts:52]: As discussed, these changes should be done  in PR #4991

File: frontend/packages/knative-plugin/src/components/overview/KnativeResourceOverviewPage.tsx [hunk 2]
```diff
@@ -42,12 +46,14 @@ export const KnativeResourceOverviewPage: React.ComponentType<KnativeResourceOve
...
   const apiInfo = groupVersionFor(item.obj.apiVersion);
-  const resourceModel = knativeModels.find(
-    (model) =>
-      model.kind === item.obj.kind &&
-      model.apiGroup === apiInfo.group &&
-      model.apiVersion === apiInfo.version,
-  );
+  const resourceModel = isDynamicEventResourceKind(referenceFor(item.obj))
+    ? getDynamicEventSourceModel(referenceFor(item.obj))
+    : knativeModels.find(
```
[the remark this answers is not in the document]
Comment (sahil143) [reply frontend/packages/knative-plugin/src/components/overview/KnativeResourceOverviewPage.tsx:52]: fixed added a condition for DynamicEventSources in stateToProps method.

File: frontend/packages/dev-console/src/components/topology/TopologyDataController.tsx [hunk 3]
```diff
@@ -116,9 +116,13 @@ export const TopologyDataController: React.FC<TopologyDataControllerProps> = ({
   resourceList,
   serviceBinding,
 }) => {
-  const { resources, utils } = getResourceList(namespace, resourceList);
+  const resourcesList = getResourceList(namespace, resourceList);
+  resourcesList.resources = [
+    ...resourcesList.resources,
+    ...getDynamicEventSourcesResourceList(namespace),
```
Comment (jeff-phillips-18) [inline frontend/packages/dev-console/src/components/topology/TopologyDataController.tsx:122]: Couldn't these be added in the plugin?

File: frontend/packages/dev-console/src/components/topology/TopologyDataController.tsx [hunk 4]
```diff
@@ -116,9 +115,9 @@ export const TopologyDataController: React.FC<TopologyDataControllerProps> = ({
   resourceList,
   serviceBinding,
 }) => {
-  const { resources, utils } = getResourceList(namespace, resourceList);
+  const resourcesList = getResourceList(namespace, resourceList);
```
Comment (invincibleJai) [inline frontend/packages/dev-console/src/components/topology/TopologyDataController.tsx:118]: Do we need the changes in this file now can have it was

File: frontend/packages/dev-console/src/utils/application-utils.ts [hunk 5]
```diff
@@ -414,6 +409,7 @@ export const cleanUpWorkload = (
   const batchDeleteRequests = (models: K8sKind[], resourceObj: K8sResourceKind): void => {
     models.forEach((model) => deleteRequest(model, resourceObj));
   };
+  if (isKnativeEventResource) deleteRequest(modelFor(referenceFor(resource)), resource);
```
Comment (invincibleJai) [inline frontend/packages/dev-console/src/utils/application-utils.ts:412]: can we have `if (isDynamicEventResourceKind(referenceFor(resource)))` ?

File: frontend/packages/dev-console/src/components/topology/TopologyDataController.tsx (hunk 3 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/topology/TopologyDataController.tsx]: fixed!

File: frontend/packages/dev-console/src/components/topology/TopologyDataController.tsx (hunk 4 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/components/topology/TopologyDataController.tsx]: fixed!

File: frontend/packages/dev-console/src/utils/application-utils.ts (hunk 5 above)
Comment (sahil143) [reply frontend/packages/dev-console/src/utils/application-utils.ts]: fixed!

**Comment by invincibleJai** [comment]
Verified locally works as expected

/lgtm

**Comment by sahil143** [comment]
@christianvogt As discussed, Raised this Jira ticket for the inconsistent ResourceIcon color on dynamic event sources. https://issues.redhat.com/browse/ODC-3587

**Pull request merged**